[RFC]trace: print core id if in atomic context

lorc opened the pull request:
If (D|E|I|F)MSG is called with foreign interrupts masked
we can report core ID. "?" will be printed instead, if
foreign interrupts aren't masked.

With this patch log looks like this:

D/TC:2 0 core_mmu_set_user_map:940 0xe181b88 0xeee8003
D/TC:? 0 __wq_rpc:40 wake  thread 1 0xe16f028 -3
D/TC:1   thread_handle_std_smc:612 a7: 2
D/TC:3 0 core_mmu_set_user_map:940 0x0 0x0

Where first digit shows core id and second - thread id.

Signed-off-by: Volodymyr Babchuk <vlad.babchuk@gmail.com>

<!--
    If you are new to submitting pull requests to OP-TEE, then please have a
    look at the list below and tick them off before submitting the pull request.

    1. Read our contribution guidelines:
         documentation/github.md.

    2. Read the contribution section in Notice.md and pay extra attention to the
       "Developer Certificate of Origin" part:
         https://github.com/OP-TEE/optee_os/blob/master/Notice.md#contributions.

    3. You should run checkpatch preferably before submitting the pull request.

    4. When everything has been reviewed, you will need to squash, rebase and
       add tags like `Reviewed-by`, `Acked-by`, `Tested-by` etc.

    NOTE: This comment will not be shown in the pull request, so no harm keeping
    it, but feel free to remove it if you like.
-->

lorc commented:
I have found that this patch is helpful when debugging SMP issues. On other hand it clogs debug output. Hence RFC tag.

jbech-linaro commented:
I'm fine with this.

jforissier commented on lib/libutils/ext/trace.c:
```diff
@@ -91,6 +117,12 @@ void trace_printf(const char *function, int line, int level, bool level_ok,
+	res = print_core_id(buf + boffs, sizeof(buf) - boffs);
+	if (res < 0)
+		return;
+	boffs += res;
```
IMO, should be conditional on `CFG_MSG_LONG_PREFIX_MASK` so that INFO messages don't show it by default (similar to thread id). And https://github.com/OP-TEE/optee_os/blob/master/mk/config.mk#L76 needs updating.

lorc commented:
Thank you. I addressed comments by @jforissier.

Also I added one more patch with small cleanup.

jforissier commented on lib/libutils/ext/trace.c:
```diff
@@ -117,39 +117,26 @@ void trace_printf(const char *function, int line, int level, bool level_ok,
-		thread_id = -1;
-	else
+		/* Print the Thread ID */
 		thread_id = trace_ext_get_thread_id();
```
For consistency with `print_core_id()` you might want to move `trace_ext_get_thread_id()` into `print_thread_id()`.

jforissier commented on lib/libutils/ext/trace.c:
```diff
@@ -91,30 +117,26 @@ void trace_printf(const char *function, int line, int level, bool level_ok,
+			return;
+		boffs += res;
 
-	if (function) {
```
You can't remove this test because it does happen that `function == NULL` (see `*MSG_RAW`) and in this case we don't want to print `(null):0`.

jforissier commented on lib/libutils/ext/trace.c:
```diff
@@ -64,6 +69,27 @@ static int print_thread_id(char *buf, size_t bs, int thread_id)
+#endif
+
+	if (thread_get_exceptions() & THREAD_EXCP_FOREIGN_INTR)
+		return snprintk(buf, bs, "%0*zi ", num_digits, get_core_pos());
```
Minor: since `get_core_pos()` returns `size_t` (unsigned), better use `%0*zu`.

lorc commented:
@jforissier, I addressed your comments.

I added a new commit that moves `trace_ext_get_thread_id()` into `print_thread_id()`.

Also I extended "check for CFG_MSG_LONG_PREFIX_MASK only once per message", because I missed an empty space.  Prior to this fix log message looked like this:

```
D/TC:3   init_secondary_helper:941 Secondary CPU Switching to normal world boot
I/TC:Dynamic shared memory is enabled
```

(note missing space between 'TC:' and 'Dynamic').

Now it looks  right:

```
D/TC:3   init_secondary_helper:941 Secondary CPU Switching to normal world boot
I/TC: Dynamic shared memory is enabled
```

jforissier commented on lib/libutils/ext/trace.c:
```diff
@@ -91,28 +117,32 @@ void trace_printf(const char *function, int line, int level, bool level_ok,
+		if (boffs >= sizeof(buf))
+		    return;
+		buf[boffs++] = ' ';
+		buf[boffs] = 0;
```
If `boffs == (MAX_PRINT_SIZE - 1)` when entering the `else {}`, this line will overflow

lorc replied:
Good catch. Thank you. I updated the whole PR, because problem code was added in fixup and I decided that fixup for fixup is overkill :)

jforissier commented:
`Reviewed-by: Jerome Forissier <jerome.forissier@linaro.org>`

lorc commented:
Thanks, @jforissier, I rebased this PR and added your r-b tag
